unitTests/UnitTest23.cpp

```diff
@@ -0,0 +1,88 @@
+
+#include <memory>
+
+#include "ElasticFrameProtocol.h"
+#include "UnitTestHelpers.h"
+
+
+// UnitTest23
+// Test sending 10 packets, each being split into six EFP fragments, five of type 1 and one of type 2
+// Drop packet 4 and 5, and deliver the rest in reversed order
+// Check that we can see that we lost two packets/superframes in the receiver when we receive the last packet
+TEST(UnitTest23, SendFivePacketsDeliverInReversedOrderWithReversedFragmentOrder) {
+    std::unique_ptr<ElasticFrameProtocolReceiver> myEFPReceiver = std::make_unique<ElasticFrameProtocolReceiver>(100,
+                                                                                                                 40);
+    std::unique_ptr<ElasticFrameProtocolSender> myEFPPacker = std::make_unique<ElasticFrameProtocolSender>(MTU);
+    std::atomic<size_t> dataReceived = 0;
+
+    size_t sentSuperFrameNumber = 0;
+    std::vector<std::vector<uint8_t>> keptBackFragments;
+    std::vector<std::vector<std::vector<uint8_t>>> keptBackSuperFrames;
+    myEFPPacker->sendCallback = [&](const std::vector<uint8_t> &subPacket, uint8_t lStreamID,
+                                    ElasticFrameProtocolContext *pCTX) {
+        if (subPacket[0] == 2) {
+            sentSuperFrameNumber++;
+            keptBackFragments.push_back(subPacket);
+            keptBackSuperFrames.push_back(keptBackFragments);
+            if (sentSuperFrameNumber == 10) {
+                for (size_t item = keptBackSuperFrames.size(); item > 0; item--) {
+                    if (item == 4 || item == 5) {
+                        continue; // Drop packet 4 and 5
+                    }
+                    std::vector<std::vector<uint8_t>> superFrame = keptBackSuperFrames[item - 1];
+                    for (size_t fragment = superFrame.size(); fragment > 0; fragment--) {
+                        ElasticFrameMessages info = myEFPReceiver->receiveFragment(superFrame[fragment - 1], 0);
+                        EXPECT_EQ(info, ElasticFrameMessages::noError);
+                    }
+                }
+            }
+            keptBackFragments.clear();
+            return;
+        }
+        keptBackFragments.push_back(subPacket);
+    };
+
+    size_t receivedFrameNumber = 0;
+    int64_t nextExpectedPts = 1001;
+    uint16_t lastReceivedSuperFrame = 0;
+    myEFPReceiver->receiveCallback = [&](ElasticFrameProtocolReceiver::pFramePtr &packet,
+                                         ElasticFrameProtocolContext *) {
+        std::cout << "Receiving frame " << packet->mPts << std::endl;
+        receivedFrameNumber++;
+        EXPECT_EQ(packet->mPts, nextExpectedPts);
+        // Expect pts 1004 and 1005 to be missing
+        nextExpectedPts += (nextExpectedPts == 1003 ? 3 : 1);
+
+        EXPECT_EQ(packet->mStreamID, 1);
+        EXPECT_EQ(packet->mCode, 0);
+        EXPECT_FALSE(packet->mBroken);
+
+        EXPECT_EQ(packet->mFrameSize, ((MTU - myEFPPacker->getType1Size()) * 5) + 12);
```

Comment:
I guess the 12 here is to match the mydata.resize() below, but is it just an arbitrary number?
Otherwise you might want to use a constant with a suitable name.

Reply:
Yes. these should match, so it makes sense to have a constant for them. I added one in all of the unit tests where I found it suitable while at it. The "12" in the expression is somewhat arbitrary, but it is chosen to force EFP to split the data into five "Type 1" fragments (which each fit `MTU - sizeof(ElasticFrameType1)` bytes of data) and one trailing type 2 fragment (which in this case is filled with the last 12 bytes)